### DISCUSSION TITLE
Create tflite model for object detection for custom dataset

### DISCUSSION THREAD

**Discussion by param-esec** [AxisCommunications/acap-computer-vision-sdk-examples#152] [General] [answered]
Hi,
Is it possible to use [TensorFlow-to-Larod-Artpec8](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/tensorflow-to-larod-artpec8) on a custom object dataset? 
Additionally, are there any other ways to train a custom object detection model using TensorFlow 1 since TensorFlow 1.x can no longer be used in Google Colab and is no longer available?
Thank you in advance.

**Comment by Corallo** [comment, marked as answer]
Hello @param-esec

Yes, you can train your own object detection. The only takeaway that you need from the TensorFlow to Larod Example is the [function](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/tensorflow-to-larod-artpec8/env/training/train.py#L73) that shows how you can quantize the model and fine-tune. You can replace the rest (model architecture and dataset) with your own.

Artpec8 DLPU supports quantization by tensor, that is only possible on TF1.
Unfortunately, this means you won't be able to train it on Colab anymore without TF1. You'll have to use either another service like AWS or train on your own GPU.

You could ignore this guideline and deploy on the device a model quantized with TF2, but you will have performance that is from 2 to 30 times worse in speed and in some cases lose also some accuracy. At that point might be better to use the CPU of the device instead.

**Reply by param-esec** [reply]
Thank you for your reply, @Corallo . It has cleared things up for me.

**Comment by param-esec** [comment]
Hi @Corallo ,
I made changes in dockerfile to add path to my custom images and annotation
![image](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/assets/137760057/1477b425-18ac-4051-9086-3ac61d336b46)

These are my classes:
"categories":[{"id":0,"name":"helmet-person","supercategory":"none"},{"id":1,"name":"Helmet","supercategory":"helmet-person"},{"id":2,"name":"Person","supercategory":"helmet-person"},{"id":3,"name":"Vest","supercategory":"helmet-person"}]

I ran this commands and got my [tflite model](https://github.com/param-esec/tfmodel/raw/main/helmet_person.tflite)
![image](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/assets/137760057/b41fd801-3a1d-4922-98b5-7580bf4cb864)

I then implemented model on [object detection example](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/tree/main/object-detector-python) but got this error:
![image](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/assets/137760057/6ed36f5d-d284-49d6-9f0d-65ed3bbe1ae1)

**Am I missing any steps for training my custom model, should i make change in other .py file?**

**Reply by Corallo** [reply]
Hello @param-esec 

Even if you have changed the dataset, you are still using the model used in the example, which is a model for binary classification, not object detection. You can see the model structure [here](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/tensorflow-to-larod-artpec8/env/training/model.py#L57).
If you want to train an object detection, you can't really use the same scripts, as also the model architecture should change, you should only use these as reference on how to do the quantization.

If you are not very practical with tensorflow, maybe you should try to follow this guide to train your object detector:
https://coral.ai/docs/edgetpu/retrain-detection/#requirements
There are also these two:
https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_detection_qat_tf1.ipynb
https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_ssdlite_mobiledet_qat_tf1.ipynb
But it might be that it is not possible anymore to use colab for TF1, you would have to reproduce them locally.

Regardless the guide that you follow, you should stop **before** exporting the model with tflite, and continue with the [fine tuning](https://github.com/AxisCommunications/acap-native-sdk-examples/blob/main/tensorflow-to-larod-artpec8/env/training/train.py#L73) of our example that does the quantization.

The error that you got in your terminal is python telling you that your model doesn't have a node called TFLite_Detection_PostProcess, as you are still using the binary image classification model.
You can look at it on netron.app

![immagine](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/assets/28478841/30d47481-9538-4091-9577-24636694f64a)
